### PR TITLE
RFC 033 Not operator

### PR DESCRIPTION
**Pull request by MacondoExpress**
# RFC 033 Not operator.

This RFC propose a generic `NOT` operator to be used as the current `AND` and `OR`  for filters.

## Complexity

Complexity: low.

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 36 | 65 | 1134 |
|  aggregations.NestedAggregation | 13514 | 13514 | 112 | 145 | 2174 |
|  aggregations.AggregationWithWhere | 13210 | 13210 | 87 | 116 | 2174 |
|  aggregations.NestedCountFromMovieToActors | 8694 | 8694 | 90 | 151 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 8900 | 8900 | 71 | 142 | 2174 |
|  aggregations.DeeplyNestedCount | 10054408 | 10054408 | 4921 | 5331 | 2008534 |
|  batch-create.BatchCreate | 3600 | 3600 | 111 | 136 | 600 |
|  connect.createAndConnect | 12419 | 12419 | 254 | 275 | 3003 |
|  connections.Connection | 13042 | 13042 | 143 | 328 | 2174 |
|  connections.NestedConnection | 33883 | 33883 | 148 | 206 | 4516 |
|  connections.ConnectionWithSort | 2277 | 2277 | 81 | 94 | 1040 |
|  connections.ConnectionWithSortAndCypher | 2277 | 2277 | 228 | 351 | 1040 |
|  create.SimpleMutation | 6 | 6 | 34 | 44 | 1 |
|  cypher-directive.TopLevelCypherDirective | 99525 | 99525 | 828 | 930 | 10168 |
|  cypher-directive.TopLevelCypherDirectiveWithColumnName | 127994 | 127994 | 216 | 338 | 10168 |
|  delete.SimpleDelete | 18361 | 18361 | 229 | 244 | 1040 |
|  delete.NestedDeleteInUpdate | 16779 | 16779 | 282 | 361 | 2040 |
|  query.SimpleQuery | 14120 | 14120 | 118 | 87 | 2174 |
|  query.QueryWhere | 12993 | 12967 | 37 | 50 | 2166 |
|  query.SimpleQueryWithNestedWhere | 13164 | 13138 | 52 | 72 | 2166 |
|  query.Nested | 10084290 | 10084290 | 9871 | 10205 | 2008534 |
|  query.NestedWithFilter | 38198 | 38198 | 129 | 148 | 4511 |
|  query.OrFilterOnRelationships | 52729 | 52311 | 177 | 227 | 2017 |
|  query.OrFilterOnRelationshipsAndNested | 39133 | 38691 | 246 | 255 | 2017 |
|  query.QueryWithNestedIn | 14820 | 14400 | 72 | 82 | 1234 |
|  query.NestedConnectionWhere | 13182 | 13182 | 56 | 74 | 2174 |
|  query.DeeplyNestedConnectionWhere | 13342 | 13236 | 89 | 110 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 6913 | 6913 | 141 | 179 | 1134 |
|  query.DeeplyNestedWithRelationshipSingleFilters | 15438 | 15438 | 102 | 118 | 2174 |
|  query.Fulltext | 96 | 96 | 30 | 36 | 16 |
|  query.FulltextWithNestedQuery | 571 | 571 | 47 | 63 | 84 |
|  sorting.SortMultipleTypes | 2493 | 2493 | 76 | 81 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 1439 | 1457 | 196 | 297 | 1040 |
|  sorting.SortOnNestedFields | 13042 | 13042 | 108 | 149 | 2174 |
|  sorting.SortDeeplyNestedFields | 38965 | 38965 | 229 | 189 | 4516 |
|  sorting.SortWithTopLevelCypher | 2119 | 2119 | 174 | 216 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 69 | 73 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 64 | 70 | 35 |
|  unions.NestedUnion | 398553 | 398553 | 319 | 338 | 33033 |
|  unions.NestedUnionWithMissingFields | 372527 | 372527 | 295 | 354 | 33033 |

</details>
        
Old Schema Generation: 33.688s
Schema Generation: 34.875s

File: docs/rfcs/rfc-034-not-operator.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+`_NOT_CONTAINS`, `_NOT_IN`, `_NOT_STARTS_WITH`.
+The goal of this RFC is to propose a mechanism to remove these extra filters in favor of a general NOT operator as it's already the case for the `AND` and `OR`.
+With the new operator is possible to reduce the schema size and have a more consistent API, currently, only a portion of predicates have the negate counterpart.
+
+## Proposed Solution
+
+This RFC proposes to deprecate all the previous filters with the `@deprecated` directive and deprecate these in the `4.x` release.
+A new operator `NOT` should then be included where the other two operators `AND` and `OR` are currently present.
+
+List of deprecated syntaxes:
```
Comment (tbwiss) [inline docs/rfcs/rfc-034-not-operator.md:17]: Just to clarify, is this the complete list of deprecated syntaxes?

File: docs/rfcs/rfc-034-not-operator.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+Almost all filters, for instance:
+`_CONTAINS`, `_IN`, `_STARTS_WITH` have their counterpart as:
+`_NOT_CONTAINS`, `_NOT_IN`, `_NOT_STARTS_WITH`.
+The goal of this RFC is to propose a mechanism to remove these extra filters in favor of a general NOT operator as it's already the case for the `AND` and `OR`.
+With the new operator is possible to reduce the schema size and have a more consistent API, currently, only a portion of predicates have the negate counterpart.
+
+## Proposed Solution
+
+This RFC proposes to deprecate all the previous filters with the `@deprecated` directive and deprecate these in the `4.x` release.
+A new operator `NOT` should then be included where the other two operators `AND` and `OR` are currently present.
```
Comment (tbwiss) [inline docs/rfcs/rfc-034-not-operator.md:15]: Are there any limitations/risks (in the code for instance) we have to keep in mind when using combinations of these three operators (NOT, AND, OR)?

**Review by tbwiss — COMMENTED** [review]
This is a helpful change!

File: docs/rfcs/rfc-034-not-operator.md (hunk 1 above)
Comment (MacondoExpress) [reply docs/rfcs/rfc-034-not-operator.md:17]: Yes, I may of course missing something!

File: docs/rfcs/rfc-034-not-operator.md (hunk 2 above)
Comment (MacondoExpress) [reply docs/rfcs/rfc-034-not-operator.md:15]: Not really, currently if we encounter an `AND` or a `OR` we wrap the following predicate using Cypher Builder: `Cypyher.and`/`Cypher.or`, to support the new operator we have just to wrap the following predicates inside a `Cypher.not`. It's already possible to chain predicates using CypherBuilder!

**Comment by MacondoExpress** [comment]
Thank you @tbwiss !

File: docs/rfcs/rfc-034-not-operator.md (hunk 1 above)
Comment (tbwiss) [reply docs/rfcs/rfc-034-not-operator.md:17]: Okay no problem!

File: docs/rfcs/rfc-034-not-operator.md (hunk 2 above)
Comment (tbwiss) [reply docs/rfcs/rfc-034-not-operator.md:15]: Perfect!

**Pull request merged**